runtime/autoload/health/provider.vim

```diff
@@ -497,6 +485,74 @@ function! s:check_for_pyenv() abort
+" sys.executable.
+function! s:locate_pythonx(invocation) abort
+  return s:normalize_path(system(a:invocation
+    \ . ' -c "import sys; sys.stdout.write(sys.executable)"'))
+endfunction
+
+" If $VIRTUAL_ENV is set, check whether its Python executables will be
+" the first on the $PATH of both Neovim and subshells spawned from
+" Neovim.
+function! s:check_active_virtualenv() abort
+  call health#report_start('Python active virtualenv')
+  let errors = []
+  " hints are kept as Dictionary keys in order to discard duplicates
+  let hints = {}
+  " the virtualenv should contain some Python executables, and those
+  " executables should be first both on Neovim's path and on the path of
+  " subshells launched from Neovim
+  let bin_dir = has('win32') ? '/Scripts' : '/bin'
+  let venv_pythonxs = glob($VIRTUAL_ENV . bin_dir . '/python*', v:true, v:true)
+  if len(venv_pythonxs)
+    for venv_pythonx in venv_pythonxs
+      let venv_pythonx = s:normalize_path(venv_pythonx)
+      let pythonx_basename = fnamemodify(venv_pythonx, ':t')
+      let neovim_pythonx = s:locate_pythonx(exepath(pythonx_basename))
+      let subshell_pythonx = s:locate_pythonx(pythonx_basename)
+      if venv_pythonx !=# neovim_pythonx
+        call add(errors, 'according to Neovim''s $PATH, the ' . pythonx_basename
+          \ . ' executable is found outside the virtualenv, here: ' . neovim_pythonx)
+        let hint = 'Problems with Neovim''s $PATH are caused by the virtualenv not being '
+          \ . 'properly activated prior to launching Neovim. Close Neovim, activate the virtualenv '
+          \ . 'properly, check that invoking Python from the command line launches the correct one, '
+          \ . 'and relaunch Neovim.'
+        let hints[hint] = v:true
+      endif
+      if venv_pythonx !=# subshell_pythonx
+        call add(errors, 'according to the $PATH of subshells launched from Neovim, the '
+          \ . pythonx_basename . ' executable is found outside the virtualenv, here: '
+          \ . subshell_pythonx)
+        let hint = 'Problems with the $PATH of subshells launched from Neovim can be '
+          \ . 'caused by your shell''s startup files overriding the $PATH previously set by the '
+          \ . 'virtualenv. Either prevent them from doing so, or use '
+          \ . 'https://vi.stackexchange.com/a/7654/18339 as a workaround.'
+        let hints[hint] = v:true
+      endif
+    endfor
+  else
+    call add(errors, 'no Python executables were found in the virtualenv''s '
+      \ . bin_dir . ' directory.')
+  endif
+
+  if len(errors)
+    call health#report_warn('$VIRTUAL_ENV is set to: ' . $VIRTUAL_ENV)
+    if len(venv_pythonxs)
+      call health#report_warn('And its ' . bin_dir . ' directory contains the following executables: '
+        \ . join(map(venv_pythonxs, "fnamemodify(v:val, ':t')"), ', '))
+    endif
+    let conj = 'However, '
+    for error in errors
+      call health#report_warn(conj . error)
+      let conj = 'And '
```

Comment:
the `conj` stuff seems overwrought. It makes the code harder to read and the resulting message might not be much better than a less clever wording.